Add TR::getMax/MinFloat/Double features

wbh123456 opened the pull request:
Add functions to get INF or -INF values of float/double.
There has been some previously defined hex values for max/min
float/double but cannot use it directly because the values are defined in
hex, e.g. [1]. The TR::getMax/MinFloat/Double functions reinterpret the hex
representations to their floating point values and return a floating
point type.

[1] https://github.com/eclipse/omr/blob/92e2fea39ad35c44d9d0c3428ea54118e4510d90/compiler/il/OMRDataTypes.hpp#L62
Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

wbh123456 commented:
@0xdaryl This PR is a supportive PR for https://github.com/eclipse/omr/pull/5105, where it requires max or min for floating point values. VPConstraints for other types (`int`/`long`/`short`) use functions such as `TR::getMaxSigned<TR::Int8 >()` to get their numeric limits. However, it might also be possible to use `std::numeric_limits()` methods to obtain the same values. Do you think whether it is a good idea to add customized `getMax/MinDouble/Float()` functions here or just utilize the `std::numeric_limits()` functions, which are already available out there?

0xdaryl commented:
I would like to say yes.  However, we've had problems in the past with `std::numeric_limits` working on all the build compilers OMR needs to support (a recent example that comes to mind is #4645).

I would say that if you're going to introduce `std::numeric_limits` for floats then I'd like to see it consistent and use the same means for all types.  You'll have to test this across all build compilers to be sure it works.  However, the experience in #4645 seems to suggest you're in for trouble.

For now, it's probably best to introduce your own limit functions like the other types have and we can revisit this when all the build compilers move up.  Perhaps open an issue so that we don't forget about this opportunity.

0xdaryl commented:
In general I'm supportive of using more from the std library.  However, if you're going to introduce `std::numeric_limits()` for floats I think you should consider making it consistent for the integer types as well.  To that end you should make sure that the answer returned by `std::numeric_limits()` matches the answer  that we presently return in our own functions (for testing purposes you could add an assert that simply verifies they return the same answer for the same type and run it on all the build compilers we know to support).  If there is a difference you need to investigate why.  I have a mild concern about performance too, perhaps influenced from the experience in #4645.

For now, so you don't lose momentum on what you're doing, I think it is probably best to introduce your own limit functions for float like the other types have (similar look-and-feel) and simply call the `std::numeric_limits` for float types under the covers.  But please open an issue so that we don't forget about this opportunity.

Leonardo2718 commented on compiler/il/OMRDataTypes.cpp:
```diff
@@ -318,3 +318,27 @@ OMR::DataType::getPrefix(TR::DataType dt)
+float TR::getMaxFloat()
+   {
+   int32_t f = FLOAT_POS_INFINITY;
+   return *(float*)&f;
```
This kind of cast relies on undefined behaviour because it violates strict aliasing.

wbh123456 replied:
It is also possible to use `reinterpret_cast` to interpret the underlining bit pattern as a floating point type[1]. Do you have any suggestion on which method I should use, or do you have better way of doing this?
[1] 
```
// x is a pointer to float inf here
int f = FLOAT_POS_INFINITY;
float *x = reinterpret_cast<float*> (&f);
```

Leonardo2718 replied:
An explicit `reinterpret_cast` doesn't change anything (the c-style cast is really just a hidden `reinterpret_cast`). The only safe way of doing this in C++ is to `memcpy` the int to a temporary float [1,2]:

[1] https://en.cppreference.com/w/cpp/string/byte/memcpy
[2]
```c++
int32_t f = FLOAT_POS_INFINITY;
float x = 0.0;
std::memcpy(&x, &f, sizeof(f));
```

wbh123456 replied:
I have used memcpy in the new commit f0a3ce4

fjeremic commented on compiler/il/OMRDataTypes.hpp:
```diff
@@ -293,6 +293,11 @@ namespace TR
    template <> inline int32_t getMaxUnsignedPrecision<TR::Int32>() { return 10; }
    template <> inline int32_t getMaxUnsignedPrecision<TR::Int64>() { return 20; }
 
+   float getMaxFloat();
+   float getMinFloat();
+   double getMaxDouble();
+   double getMinDouble();
```
These look like they should be `const` APIs. Do you agree?

wbh123456 replied:
I agree that the results should return values that are immutable and the function itself must not change any predefined variables. However, since these functions are not member of any class, it might not require the `const` keyword.

wbh123456 commented:
@0xdaryl  This PR is ready for review.

0xdaryl commented:
I am mildly concerned about the performance of these functions especially if they are going to be called frequently (say in an optimization like value propagation).  Some build compilers might be smart enough to optimize away the memcpy but I wouldn't count on all of them doing so.

The results returned are effectively constant.  Can you think of a way of memoizing the result in a static so that the mechanics of computing the constant don't occur on every call?

wbh123456 commented:
> I am mildly concerned about the performance of these functions especially if they are going to be called frequently (say in an optimization like value propagation). Some build compilers might be smart enough to optimize away the memcpy but I wouldn't count on all of them doing so.
> 
> The results returned are effectively constant. Can you think of a way of memoizing the result in a static so that the mechanics of computing the constant don't occur on every call?

I used 2 static variables to store relevant information inside each get() function. Specifically, I used a static bool variable to store if the function has been called before, and a static floating point variable to store the max/min result of memcpy. In this way, we can make sure the memcpy operation will only be performed once. However, is this implementation thread safe in OMR?

wbh123456 commented:
@0xdaryl Ping

0xdaryl commented:
I was thinking more something like this:
```
class FloatLimits
   {
   float _maxFloat;
   float _minFloat;

   void setMaxFloat()
      {
      int32_t f = FLOAT_POS_INFINITY;
      _maxFloat = 0.0f;
      memcpy(&_maxFloat, &f, sizeof(f));
      }

   void setMinFloat()
      {
      int32_t f = FLOAT_NEG_INFINITY;
      _minFloat = 0.0f;
      memcpy(&_minFloat, &f, sizeof(f));
      }

public:

   FloatLimits()
      {
      setMinFloat();
      setMaxFloat();
      }

   float getMaxFloat() { return _maxFloat; }
   float getMinFloat() { return _minFloat; }

   };




namespace TR
{

// This lives in OMRDataTypes.cpp.  The constructor of the FloatLimits class is run only once
// because this is a static instance.  The query functions below just return the cached copies
// of the float limits.

static FloatLimits floatLimits;


float getMaxFloat()
   {
   return floatLimits.getMaxFloat();
   }

float getMinFloat()
   {
   return floatLimits.getMinFloat();
   }

}
```

0xdaryl commented:
Yes, that's more like what I had in mind.  See your compilation errors though.  I think I would rather fix this by moving the `static FloatLimits` into OMRDataTypes.cpp and the getter functions as well as externs.  That way, you avoid creating new functions in each file you include OMRDataTypes.hpp into  (we've had footprint problems in the past when that was done for some bigger functions).

wbh123456 commented:
> Yes, that's more like what I had in mind. See your compilation errors though. I think I would rather fix this by moving the `static FloatLimits` into OMRDataTypes.cpp and the getter functions as well but as externs. That way, you avoid creating new functions in each file you include OMRDataTypes.hpp into (we've had footprint problems in the past when that was done for some bigger functions).

I have fixed the linker error in 431c89b. I moved the definitions of getter functions to .cpp file to avoid multiple definitions. Also make `static FloatLimits fpLimits` to `extern FloatLimits fpLimits` and create a definition of it in .cpp file. In this way, I think the `FloatLimits` is constructed exactly once, any call to the getter functions will look into `_maxFloat`/`_minFloat` that are cached in the fpLimits instance.

0xdaryl commented:
I'm not quite following why you don't make `FloatingPointLimits` a static in the cpp file.  In addition to the initialization benefits, by making it static you also hide it (encapsulate it) within that compilation unit as it arguably is only useful in that file and shouldn't be exposed more broadly.

0xdaryl commented:
@genie-omr build all
